### DISCUSSION TITLE
How to find function arguments for non documented modules? (for the purpose of trying to write a cross port library)

### DISCUSSION THREAD

**Discussion by Always-Self-Hosted** [micropython/micropython#14058] [Using MicroPython]
I've been going down the road of trying to write a library that will work across the esp32, esp8266 and rp2 ports, and so I'm only using modules that are available on all of those boards for a given micropython version, which I've been getting by using the REPL to `help("modules")` for each board. 

But trying to go further with that by only using functions that are shared across all ports that take the same amount of arguments has me a bit stuck. The micropython documentation site is great for all of the info on there, but as soon as i want something that isn't documented, an arbitrary example of that being DHT11 (actually probably not the BEST example since it does have BASIC documentation under the esp reference with an example, but not for rp2 and not "proper" documentation"), I can't seem to find a good way to ensure if writing portable code using that module is possible. 

I initially tried grabbing the stubs myself off each board, which was time consuming, and ended up resulting in function arguments not being available anyway, and i read online that others in the python community use the  `inspect` module to do this, but that does not seem to be available on micropython, likewise with many dunder methods on objects. I've also seen some micropython specific info online about getting details from reading the c code, and if that's a genuinely possibility (meaning you can READ the differences between different ports), then i can live with that, but i would need help as i did try that, but I'm not familiar with the C language, and i couldn't even find a specific function i wanted to know about trying to navigate the repo naively.  

So I'm looking for direction or advice on how to find if a specific function is available on a port, and which arguments it takes for that port. How are others writing microptyhon code to be portable across ports? How do you develop with micropython efficiently without relying on the documentation website for every function argument, or when there is no documentation available at all?

**Comment by Josverl** [comment]
I started out with a similar question about 5 years ago, and based on that I created the micropython-stubs project to combine the available information from available sources ( docs , frozen .py, interrogating the firmware directly) , and make that available to IDEs and typecheckers.
So you can have that at your fingertips in most IDEs if you so want. But only for one port/board at a time.

I agree that there are still a large number of MicroPython classes, methods and functions that have very sparse, or no documentation.
I have been tracking this across versions, and while the percentage of documented modules/classes and methods is steadily growing , but still not there.

And if there is no documentation - there is no information on the parameters. 
Ive tried to build a parser for the C(++) modules - but that yielded too little usable information , partly because several different coding styles 
and patters are used by the authors of different modules and ports.

As MicroPython tries to be as similar to CPython as possible, its documentation is a good fallback , and this is why i include a link to the CPython stdlib docs where possible.

On comparing the similarities/differences across ports, I know there are many and I have some thoughts on how to compare those  ( Run a AST / CST parser on the stubs, and compare the syntax trees ) 
However in order to make the result useful, I would need to find a way to visualize that at scale ( 150+ boards x ~60 modules x ~25 methods/classes/functions ) x the number of versions.
Putting that into a database and something like PowerBI for querying and slicing and dicing should be doable.
I dabbed a bit in looking for an web hosted / OSS frontend, and for me that would take months compared to hours in PowerBI, so i do not think that is a good use of my time.

Update:
I could try to build / merge a sort of micropython-allup-stubs, where I merge all stubs and also record the origin of the information.
I think the big challenge again will be to keep that readable, but at least that will reduce the problem by a few dimensions.

**Reply by Always-Self-Hosted** [reply]
First of all, i very much appreciate the time taken for the response and the detail given for things you've already thought through. 

I guess it leaves me in the same spot i was in though. Is there something tangible to do for a novice to progress any of the ideas you had to solve the issues in this space? Or is this just a hard problem that can only be solved with real engineering hours?

**Reply by Josverl** [reply]
Circuitpython has put great efforts into a a single and unified abstraction layer, precisely because novice programmers should not need to care about differences. 
The language is the same as MicroPython- however the (hardware) APIs and libraries are different.

Just to be clear: I have no personal experience with that, but I've heard good things

**Reply by Josverl** [reply]
**One last thing:**
I have created this [overview of what modules exists on which ports / boards and versions](https://flatgithub.com/Josverl/micropython-stubs/?filename=all_modules.json)
and this allows one to search for all ports/ bords that have a specific module such as [`aioble.central`](https://flatgithub.com/Josverl/micropython-stubs/?filename=all_modules.json&filters=mod_name%3Dcentral&sha=4a7c5bbbafa92ae8cd200a4afdf36ebf1ed0ef5a&sort=family%2Casc&stickyColumnName=family)
the list is a .json document which is updated automatically based on the stubs that are generated.

Would it help if I add classes methods and their parameters to that to allow lookup / comparison across the different ports and boards ?
Note that it still wont have the undocumented information 😕
